File: c++/leetcode/513. Find Bottom Left Tree Value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  int findBottomLeftValue(TreeNode *root)
  {
    queue<TreeNode *> q;
    q.emplace(root);

    int ans{};

    while (!q.empty())
    {
      const int sz = q.size();

      for (int i = 0; i < sz; ++i)
      {
        const auto front = q.front();

        if (front->right)
          q.emplace(front->right);
        if (front->left)
          q.emplace(front->left);
        ans = front->val;
        q.pop();
      }
    }

    return ans;
  }
};

File: javascript/leetcode/513. Find Bottom Left Tree Value.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findBottomLeftValue = function (root) {
  let q = [root];
  let ans = root.val;

  while (q.length) {
    const sz = q.length;

    for (let i = 0; i < sz; ++i) {
      const front = q[0];

      if (front.right) q.push(front.right);
      if (front.left) q.push(front.left);

      ans = front.val;
      q = q.slice(1);
    }
  }

  return ans;
};
